$MirOS: ports/net/centericq/patches/patch-src_hooks_abstracthook_cc,v 1.2 2007/12/23 23:27:28 bsiegert Exp $
--- src/hooks/abstracthook.cc.orig	Sun Aug 17 00:39:03 2008
+++ src/hooks/abstracthook.cc	Sun Aug 17 00:52:22 2008
@@ -42,11 +42,24 @@
 
 #define NOTIFBUF 512
 
+extern "C" char *sautoconv_c(const char *);
+
+static string sautoconv(const string &atext) {
+	const char *ccp;
+	string r;
+
+	ccp = sautoconv_c(atext.c_str());
+	r = ccp;
+	return r;
+}
+
 time_t timer_current = time(0);
 
 abstracthook::abstracthook(protocolname aproto)
     : proto(aproto), searchdest(0)
-{ }
+{
+	protocharset = NULL;
+}
 
 void abstracthook::init() {
 }
@@ -233,6 +246,19 @@ string abstracthook::ruscrlfconv(const s
 
 string abstracthook::rusconv(const string &tdir, const string &text) {
     string r;
+
+    if (conf.getisunicode()) {
+#ifdef HAVE_ICONV
+	if (protocharset != NULL) {
+		if (tdir == "kw" || tdir == "ku")
+			return siconv(text, "utf-8", protocharset);
+		if (tdir == "wk" || tdir == "uk")
+			return siconv(text, protocharset, "utf-8");
+		return text;
+	}
+#endif
+	return sautoconv(text);
+    }
 
     if(!conf.getcpconvert(proto) && tdir.find("u") == -1)
 	return text;
